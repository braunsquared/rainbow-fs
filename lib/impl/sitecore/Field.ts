import { IField } from '../../model/Sitecore';

export class Field implements IField {
  ID: string = "";
  Hint: string;
  Type?: string;
  Value: any;

  constructor(id: string, hint: string, value: any) {
    this.ID = id;
    this.Hint = hint;
    this.Value = value;
  }

  toObject(): any {
    const obj: any = {
      ID: this.ID,
      Hint: this.Hint,
      Value: this.Value,
    };

    if(this.Type) {
      obj.Type = this.Type;
    }

    return obj;
  }

  static fromObject(obj: any): IField {
    const field = new Field(obj.ID, obj.Hint, obj.Value);
    field.Type = obj.Type;
    return field;
  }
}