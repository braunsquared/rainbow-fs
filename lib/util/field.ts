import { GUID, IField } from '../model/Sitecore';
import { toGUID } from './guid';

export function treelistValue(field: IField | undefined, defaultValue: GUID[] = []): GUID[] {
  if (!field || !field.Value) {
    return defaultValue;
  }

  return field.Value.split(/\n|\r\n/).map(id => toGUID(id)).filter(i => i);
}

export function booleanValue(field: IField | undefined, defaultValue: boolean = false): boolean {
  if (!field || !field.Value) {
    return defaultValue;
  }

  return field.Value === '1';
}