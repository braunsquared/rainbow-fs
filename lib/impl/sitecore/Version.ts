import { IVersion, IField } from '../../model/Sitecore';
import { Field } from "./Field";

export class Version implements IVersion {
  Version: number = 1;
  Fields: Array<IField> = [];

  constructor(version: number) {
    this.Version = version;
  }

  addField(field: IField): this {
    this.Fields.push(field);
    return this;
  }

  removeField(field: IField): this {
    const idx = this.Fields.indexOf(field);
    if (idx >= 0) {
      this.Fields.splice(idx, 1);
    }
    return this;
  }

  field(idOrName: string): IField | undefined {
    return this.Fields.find(f => f.ID === idOrName || f.Hint === idOrName);
  }

  toObject(): any {
    const obj: any = {
      Version: this.Version
    };

    if (this.Fields.length) {
      obj.Fields = this.Fields.map(f => f.toObject());
    }

    return obj;
  }

  static fromObject(obj: any): Version {
    const version = new Version(obj.Version);

    if (obj.Fields) {
      version.Fields = obj.Fields.map((fieldObj: any) => Field.fromObject(fieldObj));
    }

    return version;
  }
}