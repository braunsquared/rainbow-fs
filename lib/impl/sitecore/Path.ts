import * as path from '../../util/path';
import { IPath } from '../../model/Sitecore';

export class Path implements IPath {
  readonly Name: string;
  readonly Folder: string;
  readonly Path: string;
  readonly Tokens: Array<string>;

  constructor(scPath: string) {
    const parsed = path.parse(scPath);
    this.Name = parsed.name;
    this.Folder = parsed.base;
    this.Path = parsed.path;
    this.Tokens = parsed.tokens;
  }

  rebase(oldRoot: string, newRoot: string): Path {
    // TODO: handle the fact that sitecore is not case sensitive
    if (!this.Path.startsWith(oldRoot)) {
      throw new Error('Unable to rebase when old root does not match');
    }

    return new Path(`${newRoot}/${this.Path.substring(oldRoot.length)}`);
  }

  relativePath(childPath: IPath): Path {
    if (!childPath.Path.startsWith(this.Path)) {
      throw new Error('Path is not a base of the child path');
    }

    return new Path(childPath.Path.substring(this.Path.length));
  }
}