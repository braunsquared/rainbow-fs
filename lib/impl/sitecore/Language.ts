import { ILanguage, IVersion, IField } from '../../model/Sitecore';
import { Version } from "./Version";
import { Field } from './Field';

export class Language implements ILanguage {
  Language: string = "en";
  Fields: Array<IField> = [];
  Versions: Array<IVersion> = [];

  constructor(lang: string) {
    this.Language = lang;
  }

  addVersion(v: IVersion): this {
    this.Versions.push(v);
    return this;
  }

  removeVersion(v: IVersion): this {
    const idx = this.Versions.indexOf(v);
    if (idx >= 0) {
      this.Versions.splice(idx, 1);
    }
    return this;
  }

  latestVersion(): IVersion | undefined {
    let latest : IVersion = this.Versions[0];

    this.Versions.forEach(v => {
      if (!latest || (latest && latest.Version < v.Version)) {
        latest = v;
      }
    });

    return latest;
  }

  addField(field: IField): this {
    this.Fields.push(field);
    return this;
  }

  removeField(field: IField): this {
    const idx = this.Fields.indexOf(field);
    if (idx >= 0) {
      this.Fields.splice(idx, 1);
    }
    return this;
  }

  field(idOrName: string): IField | undefined {
    const field = this.Fields.find(f => f.ID === idOrName || f.Hint === idOrName);
    
    if(field) {
      return field;
    }

    const version = this.latestVersion();
    if (version) {
      return version.field(idOrName);
    }
  }

  toObject(): any {
    const obj: any = {
      Language: this.Language,
    };

    if (this.Fields.length) {
      obj.Fields = this.Fields.map(f => f.toObject());
    }

    if (this.Versions.length) {
      obj.Versions = this.Versions.map(v => v.toObject());
    }

    return obj;
  }

  static fromObject(obj: any): Language {
    const lang = new Language(obj.Language);

    if (obj.Fields) {
      lang.Fields = obj.Fields.map((fObj: any) => Field.fromObject(fObj));
    }

    if (obj.Versions) {
      lang.Versions = obj.Versions.map((vObj: any) => Version.fromObject(vObj));
    }

    return lang;
  }
}