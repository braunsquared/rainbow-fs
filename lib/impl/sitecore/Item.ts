import { GUID, IItem, IPath, ILanguage, IField, ItemMetaData } from '../../model/Sitecore';
import { Field } from './Field';
import { Language } from './Language';
import { Path } from './Path';
import { IStore } from '../../Store';

export function defaultItemFactory(store: IStore, obj: any, meta?: ItemMetaData): IItem {
  const itemObj = {...obj};
  if (!itemObj.ID) {
    itemObj.ID = store.generateId(store.config.idSource(itemObj));
  }

  const item = Item.fromObject(itemObj);

  if(meta) {
    item.Meta = {...meta};
  }

  return item;
}

export class Item implements IItem {
  ID: GUID;
  Parent: GUID;
  Template: GUID;
  Path: IPath;
  DB: string = "master";

  SharedFields: Array<IField> = [];
  Languages: Array<ILanguage> = [];

  Meta: ItemMetaData = {
    isNew: true,
    isDirty: false,
    isVirtual: false,
  }

  constructor(id: GUID, path: string) {
    this.ID = id;
    this.Path = new Path(path);
  }

  addSharedField(field: IField): this {
    this.SharedFields.push(field);
    return this;
  }

  removeSharedField(field: IField): this {
    const idx = this.SharedFields.indexOf(field);
    if (idx >= 0) {
      this.SharedFields.splice(idx, 1);
    }
    return this;
  }

  addLanguage(lang: ILanguage): this {
    this.Languages.push(lang);
    return this;
  }

  removeLanguage(lang: ILanguage): this {
    const idx = this.Languages.indexOf(lang);
    if (idx >= 0) {
      this.Languages.splice(idx, 1);
    }
    return this;
  }

  field(idOrName: string, lang: string = 'en'): IField | undefined {
    const field = this.SharedFields.find(f => f.ID === idOrName || f.Hint === idOrName);
    
    if (field) {
      return field;
    }

    const langInst = this.Languages.find(l => l.Language === lang);
    if (langInst) {
      return langInst.field(idOrName);
    }
  }

  toObject(): any {
    const obj: any = {
      ID: this.ID,
      Parent: this.Parent,
      Template: this.Template,
      Path: this.Path.Path,
      DB: this.DB,
    };

    if (this.SharedFields.length) {
      obj.SharedFields = this.SharedFields.map(f => f.toObject());
    }

    if (this.Languages.length) {
      obj.Languages = this.Languages.map(l => l.toObject());
    }

    return obj;
  }

  static fromObject(obj: any): Item {
    const item = new Item(obj.ID, obj.Path);
    item.Parent = obj.Parent;
    item.Template = obj.Template;
    item.DB = obj.DB;

    if (obj.SharedFields) {
      item.SharedFields = obj.SharedFields.map((fieldObj: any) => Field.fromObject(fieldObj));
    }

    if (obj.Languages) {
      item.Languages = obj.Languages.map((langObj: any) => Language.fromObject(langObj));
    }

    return item;
  }
}