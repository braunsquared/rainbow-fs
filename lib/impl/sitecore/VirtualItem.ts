import { Item } from './Item';
import { IField, ILanguage } from '../../model/Sitecore';

export class VirtualItem extends Item {
  Meta = {
    isNew: false,
    isDirty: false,
    isVirtual: true,
  }

  addSharedField(field: IField): this {
    throw new Error('Unable to perform action on virtual item.');
  }

  removeSharedField(field: IField): this {
    throw new Error('Unable to perform action on virtual item.');
  }

  addLanguage(lang: ILanguage): this {
    throw new Error('Unable to perform action on virtual item.');
  }

  removeLanguage(lang: ILanguage): this {
    throw new Error('Unable to perform action on virtual item.');
  }
}