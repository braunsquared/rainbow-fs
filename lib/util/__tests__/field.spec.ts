import { Field } from '../../impl/sitecore/Field';
import { treelistValue } from '../field';

describe('field util', () => {
  it('should parse a treelist value with a single item', () => {
    const field = new Field('abc', '', '{1930BBEB-7805-471A-A3BE-4858AC7CF696}');
    expect(treelistValue(field)).toMatchSnapshot();
  });

  it('should parse a treelist value with multiple item', () => {
    const field = new Field('abc', '', '{1930BBEB-7805-471A-A3BE-4858AC7CF696}\n{B9764823-CF16-4E03-BEAA-C0EABED31D6C}\n');
    expect(treelistValue(field)).toMatchSnapshot();
  });
});