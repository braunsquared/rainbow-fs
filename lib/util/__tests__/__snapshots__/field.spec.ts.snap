// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`field util should parse a treelist value with a single item 1`] = `
Array [
  "1930bbeb-7805-471a-a3be-4858ac7cf696",
]
`;

exports[`field util should parse a treelist value with multiple item 1`] = `
Array [
  "1930bbeb-7805-471a-a3be-4858ac7cf696",
  "b9764823-cf16-4e03-beaa-c0eabed31d6c",
]
`;
